Lib/test/test_exceptions.py

```diff
@@ -1330,6 +1331,40 @@ def test_assert_shadowing(self):
             del AssertionError
             self.fail('Expected exception')
 
+    def test_memory_error_subclasses(self):
+        # MemoryError instances use a freelist of objects that are
+        # linked using the 'dict' attribute when they are
+        # inactive/dead. Subclasses of MemoryError should not
+        # participate in the freelist schema. This test creates a
+        # MemoryError object and keeps it alive (therefore
+        # advancing the freelist) and then it creates and destroys
+        # a subclass object. Finally, it checks that creating a
+        # new MemoryError succeeds, proving that the freelist is
+        # not corrupted. See bpo-41654 for more information.
+
+        Check bpo-41654 for more information.
+        class TestException(MemoryError):
+            pass
+
+        try:
+            raise MemoryError
+        except MemoryError as exc:
+            inst = exc
+
+        try:
+            raise TestException
+        except Exception:
+            pass
+
+        inst = None
+
+        try:
+            raise MemoryError
+        except MemoryError as exc:
+            pass
+
+        gc.collect()
```

Comment:
On my machine (on a release build), the proposed test is giving a false negative and passing before this change.

It crashes only when repeated, as in:
```Python
import gc

class TestException(MemoryError):
    pass

for _ in range(10):
    try:
        raise MemoryError
    except MemoryError as exc:
        inst = exc

    try:
        raise TestException
    except Exception:
        pass

    gc.collect()
```

Reply:
That's expected because the corruption is only detected immediately in debug mode.

Reply:
In any case, I have pushed a new amend to run the last part (which is the one that collects the corruption created by `TestException` in a loop. This triggers on release builds as well.